Hypothalamus Implementation

santiestrada32 opened the pull request:
New

- Inclusion of Hypothalamus module
- Inclusion of the hypothalamus labels into FastSurfer LUT

TODO

- Include HypVINN in fastsurfer segmentation pipeline
- Include T2 input. The input T2 image should be biased field corrected and co-register to the T1 (see HypVINN readme file)

santiestrada32 committed:
HypVINN implementation
without __pycache__ files
include HypVINN labels
Create Main function for N4 bias correct
include fs bin for registration

santiestrada32 committed:
Include Copyright statements

Include copyright statements in Hypvinn files

santiestrada32 committed:
assert input images

santiestrada32 committed:
Update run_pipeline.py

santiestrada32 committed:
update visualization
update run pipeline
test fs

santiestrada32 committed:
include readmefile and remove test scripts

santiestrada32 committed:
add documentation

santiestrada32 commented:
I was getting an attribute error with the current import of the yacs library
Error:
AttributeError: module 'yacs' has no attribute 'config'

dkuegler commented:
> I was getting an attribute error with the current import of the yacs library Error: AttributeError: module 'yacs' has no attribute 'config'

That looks like a typing bug.
Just put quotes around it in the typing for now

dkuegler commented on Docker/install_fs_pruned.sh:
```diff
@@ -372,7 +374,6 @@ link_files="
   bin/mri_surf2vol
   bin/mri_surfcluster
-  bin/mri_vol2vol
   bin/mri_voldiff
```
We probably want to change that back as it was alphabetically sorted before

dkuegler replied:
also is mri_robust_register in the list?

dkuegler commented on HypVINN/config/HypVINN_axial_v1.0.0.yaml:
```diff
@@ -0,0 +1,91 @@
+{
```
This file looks like it is Json formatted, not yaml. Is that on purpose?

dkuegler commented on HypVINN/config/HypVINN_coronal_v1.0.0.yaml:
```diff
@@ -0,0 +1,91 @@
+{
```
Again Json format, yaml file extension

dkuegler commented on HypVINN/utils/mode_config.py:
```diff
@@ -0,0 +1,101 @@
+    
+    if t1_mode:
+        assert os.path.isfile(t1_path), f"T1 image not found"
+        t1 = nib.load(t1_path)
```
I don't know if we have to fix this right now, but if you are loading the header here anyways, why not just call it a load images function that asserts, but also returns the image objects? This way we don't have to wait for the io 2 extra times?

dkuegler commented on HypVINN/README.md:
```diff
@@ -0,0 +1,123 @@
+* Zenodo records: https://zenodo.org/records/10623893
+
+### Pipeline Steps
+1. Bias Field Correction Step
```
This should be optional (run_fastsurfer has already run bias field correction at this point). Mayb we need to check that the intensity ranges are not too different because of this, but I see no real argument to run bias field correction several times in the pipeline...
This is pretty important to fix soon-ish.

dkuegler replied:
I guess you can specify --no_bc, we still need to check if our bias field works well with hypo-seg, because the intensities might be different....

dkuegler commented on HypVINN/README.md:
```diff
@@ -0,0 +1,123 @@
+ * `--sd <name>` : Directory in which evaluation results should be written.
+ *  `--t1 </dir/T1**.nii.gz>` : T1 image path, required = True
+ *  `--t2 </dir/T2**.nii.gz>` : T2 image path, T2 image not required when running the t1 mode for the others is required , required = False
+ * `--mode <name>` : Mode to run segmentation based on the available modalities. If is set to auto the model will choose the mode based on the passed input images, 
```
Is the mode not implicit by which files are passed? I mean, I cannot run t1 and t2 mode if I only pass t1 images, right?

dkuegler commented on HypVINN/README.md:
```diff
@@ -0,0 +1,123 @@
+                     t1 : only T1 images, t2 : only T2 images or multi : both T1 and T2 images, default = 'auto'
+ * `--seg_log` :  Path to file in which run logs will be saved. If not set logs will be stored in /sd/sid/logs/hypvinn_seg.log 
+### Image processing options
+ * `--no_pre_proc`: Deactivate all pre-processing steps, This is recommended when images are already bias field corrected and co-register if T1 and T2 are available
```
Is this --no_bc + --no_reg?

dkuegler commented on HypVINN/README.md:
```diff
@@ -0,0 +1,123 @@
+ * `--no_pre_proc`: Deactivate all pre-processing steps, This is recommended when images are already bias field corrected and co-register if T1 and T2 are available
+ * `--no_bc` : Deactivate bias field correction, it is recommended to do bias field correction for calculating volumes taking account partial volume effects, reguired = False
+ * `--no_reg` : Deactivate registration of T2 to T1. If multi mode is used; images need to be register properly, required = False
+ * `--reg_type` : Freesurfer Registration type to run. coreg : mri_coreg (Default) or robust : mri_robust_register.
```
I am not sure whether mri_robust_register is in install_fs_pruned.

dkuegler commented on HypVINN/README.md:
```diff
@@ -0,0 +1,123 @@
+            |-- hypothalamus_mask.HypVINN.nii.gz (Hypothalamus Segmentation Mask)
+            |-- transforms
+                |-- t2tot1.lta (FreeSurfer registration file, only available if registration is performed)
+        |-- qc_snapshots : QC outputs
```
I think there should be a flag to either activate or deactivate this output. Maybe we discuss this.

dkuegler commented on HypVINN/config/HypVINN_ColorLUT.txt:
```diff
@@ -0,0 +1,26 @@
+#No. Label Name:                R   G   B   A
```
Is this not basically the same as the FreeSurfer_ColorLUT.txt (or rather all these are also in the FS_ColorLUT)?

I think this file maybe should have the "internal" class numbers, so the class output ids of the network?

dkuegler commented on HypVINN/config/hypvinn_global_var.py:
```diff
@@ -0,0 +1,139 @@
+# limitations under the License.
+
+import numpy as np
+HYPVINN_CLASS_NAMES = {
```
Should this "internal" mapping not be the hypvinncolorlut?

dkuegler commented on HypVINN/README.md:
```diff
@@ -0,0 +1,123 @@
+
+### Requirements
+Same as FastSurfer
+FreeSurfer should also be source to run the code and the mri_coreg ang mri_vol2vol binary should be also available
```
Can you add code that it is enough if FREESURFER_HOME is defined?

dkuegler committed:
Update README.md

Some typos and clarification

dkuegler commented on .gitignore:
```diff
@@ -1,2 +1,12 @@
+#Dont Commit checkpoints
+checkpoints/
+
+HypVINN/test
```
I am not sure about adding HypVINN/test here... the rest I am not sure about why it was not added before...

dkuegler commented on HypVINN/data_loader/data_utils.py:
```diff
@@ -0,0 +1,266 @@
+    return img
+
+# Transformation for mapping
+def transform_axial2coronal(vol, axial2coronal=True):
```
This is pretty confusing with respect to `FastSurferCNN.data_loader.data_utils.transform_axial`.

Please make sure you comment why this needs to be different, e.g. this is assuming RAS instead of LIA orientation. Otherwise this will cause dumb errors.

dkuegler commented on HypVINN/data_loader/data_utils.py:
```diff
@@ -0,0 +1,266 @@
+    else:
+        return np.moveaxis(vol, [0, 1, 2], [0, 2, 1])
+
+def transform_axial2sagittal(vol, axial2sagittal=True):
```
As above.

dkuegler commented on HypVINN/data_loader/data_utils.py:
```diff
@@ -0,0 +1,266 @@
+    new_data = np.uint8(np.rint(mapped_data))
+    return new_data
+
+def getscale(data, dst_min, dst_max, f_low=0.0, f_high=0.999):
```
Pretty sure this is the same as FastSurferCNN.data_loader.conform.getscale

dkuegler commented on HypVINN/data_loader/data_utils.py:
```diff
@@ -0,0 +1,266 @@
+    # print("rescale:  min: " + format(src_min) + "  max: " + format(src_max) + "  scale: " + format(scale))
+    return src_min, scale
+
+def scalecrop(data, dst_min, dst_max, src_min, scale):
```
Pretty sure this is identical to FastSurferCNN.data_loader.conform.scalecrop

dkuegler commented on HypVINN/data_loader/data_utils.py:
```diff
@@ -0,0 +1,266 @@
+    else:
+        return np.moveaxis(vol, [0, 1, 2], [1, 2, 0])
+
+def rescale_image(img_data):
```
maybe add a note: Same as CerebNet.datasets.utils.rescale_image

dkuegler commented on HypVINN/inference.py:
```diff
@@ -0,0 +1,186 @@
+        logger.info(f"Running view aggregation on {self.viewagg_device}")
+
+        # Options for parallel run
+        self.model_parallel = (
```
I dont really know if it makes a lot of sense to do model_parallel in inference. I'd just always do single-gpu inference. If people have multiple GPUs, they should do different subjects on different GPUs...

dkuegler commented on HypVINN/inference.py:
```diff
@@ -0,0 +1,186 @@
+        model = build_model(self.cfg)  #
+        model.to(self.device)
+
+        if self.model_parallel:
```
see above

dkuegler commented on HypVINN/inference.py:
```diff
@@ -0,0 +1,186 @@
+
+    def run(self, subject_name, modalities, orig_zoom, pred_prob, out_res=None,mode='multi'):
+        # Set up DataLoader
+        test_dataset = HypoVINN_dataset(subject_name, modalities, orig_zoom, self.cfg, mode = mode,
```
As we have it here right now, the image will be read from disk three times.

I don't know if this needs to be fixed right in this initial commit, but CerebNet provides the pattern to fix this.
If this is fixed, it probably also makes sense to fix the RAS/LIA loading (again similar to how it is done in CerebNet).

dkuegler commented on HypVINN/data_loader/dataset.py:
```diff
@@ -0,0 +1,115 @@
+        else:
+            logger.info(f"For inference T1 block weight was set to : {self.weight_factor.numpy()[0]} and the T2 block was set to: {self.weight_factor.numpy()[1]}")
+
+    def _conform_img(self,orig_data,orig_zoom,modalitie):
```
Can we rename this function?
conforming means reorient to LIA, robust intensity scaling etc.

dkuegler commented on HypVINN/models/networks.py:
```diff
@@ -0,0 +1,191 @@
+        skip_encoder_02 = self.m2_inp_block(x[1])
+
+
+        s_weights = self.smooth_weights(self.mod_weights)
```
Maybe rename smooth_weights to normalize_weights?

dkuegler commented on HypVINN/models/networks.py:
```diff
@@ -0,0 +1,191 @@
+        """
+        eps = 0.00000001
+        # Comparison weights to set channels to 1 or zero
+        comparison = weight_factor[0]
```
We definitely need some documentation/typing on the weight_factor.

dkuegler replied:
especially, since s_weights, which the relevant number only derives from mod_weights, which is a parameter...
why is weight_factor not a `modality_mask`with data type `bool`?

santiestrada32 replied:
i added more documentation

dkuegler commented on HypVINN/models/networks.py:
```diff
@@ -0,0 +1,191 @@
+        if 1 in comparison:
+            s_weights = s_weights * comparison
+            s_max,_ = torch.max(s_weights,dim=0)
+            s_weights = s_weights / (s_max+eps)
```
This is a bit surprising to me. Why not just use `torch.Tensor(comparison, dtype=float)`? that would be both simpler, more easy to understand and error-robust...

dkuegler commented on HypVINN/run_pipeline.py:
```diff
@@ -0,0 +1,150 @@
+# Copyright 2024 AI in Medical Imaging, German Center for Neurodegenerative Diseases(DZNE), Bonn
+#
```
I think you should merge run_pipeline and run_prediction into one file, i.e. move all the code from run_pipeline into run_prediction. Then it would be a bit more similar to FastSurferCNN and CerebNet

dkuegler commented on HypVINN/utils/N4_bias_correct.py:
```diff
@@ -0,0 +1,309 @@
+#!/usr/bin/env python3
+
+
+# Copyright 2019 Image Analysis Lab, German Center for Neurodegenerative Diseases (DZNE), Bonn
```
I think we should look at whether it is possible to just use `recon_surf/N4_bias_correct.py` here. I am not really sure what you changed in the script. Maybe you can say what the difference is?

dkuegler replied:
We need to discuss if we want the N4_bias_correct to even stay in HypVINN (or basically force --no_bc always)

dkuegler commented on HypVINN/README.md:
```diff
@@ -0,0 +1,123 @@
+* Zenodo data repository: https://zenodo.org/records/10623893
+
+### Pipeline Steps
+1. Bias Field Correction Step (optional)
```
The more I think about it the more I am thinking that bias field correction should not be part of your script. We can then remove all the bias correction code from HypVINN, dont have the scripts in there multiple times etc.
In the standard (non-developer) way to run HypVINN (and FastSurfer), everything (including the biasfield) is orchestrated via run_fastsurfer.sh, so we KNOW we have a bias-field corrected image. And we just write into the API for HypVINN, that the t1 image needs to be bias-field corrected.
This is also kind of how CerebNet does it (the bias field corrected image is only used for segstats there, but still).

dkuegler commented on HypVINN/utils/img_processing_utils.py:
```diff
@@ -0,0 +1,261 @@
+    return clean_seg,labels_cc, savemask
+
+
+def run_N4_bias_correct(invol,outvol,threads=1):
```
maybe get rid of this function

dkuegler commented on HypVINN/utils/img_processing_utils.py:
```diff
@@ -0,0 +1,261 @@
+
+    return flag
+
+def N4_bias_correct(t1_path,t2_path,mode,out_dir,threads=1):
```
maybe get rid of this function

dkuegler commented on HypVINN/run_preproc.py:
```diff
@@ -0,0 +1,54 @@
+# Copyright 2024 AI in Medical Imaging, German Center for Neurodegenerative Diseases(DZNE), Bonn
```
I would also move this into a different file or rename this file to `preproc.py`.

The FastSurfer convention right now is:
if the file starts with *run_* it is a script, if not, it may be a script, but it also is sometimes a library.

dkuegler commented on HypVINN/utils/misc.py:
```diff
@@ -0,0 +1,17 @@
+import shlex
+
+
+def run_cmd(cmd):
```
remove this file, see img_processing_utils as well.

dkuegler commented on HypVINN/utils/img_processing_utils.py:
```diff
@@ -0,0 +1,261 @@
+    return t1_bc_path,t2_bc_path
+
+
+def t1_to_t2_registration(t1_path,t2_path,out_dir,registration_type='coreg',bc_status=False):
```
I am not 100% convinced we want to keep this function as is. Maybe we should just move the call to run_fastsurfer as well...

dkuegler commented on HypVINN/utils/img_processing_utils.py:
```diff
@@ -0,0 +1,261 @@
+    if registration_type == 'coreg':
+        cmd = 'mri_coreg --mov {} --targ {} --reg {}'.format(t2_path, t1_path, lta_path)
+        LOGGER.info(cmd)
+        run_cmd(cmd)
```
Instead of `run_cmd` I would use `Popen` from `FastSurferCNN.utils.run_tools`.

```python
        cmd = 'mri_coreg --mov {} --targ {} --reg {}'.format(t2_path, t1_path, lta_path)
        LOGGER.info(cmd)
        run_cmd(cmd)
```
        
   
```python
        exe = shutil.which("mri_coreg")
        if not bool(exe):
             if os.environ.get("FREESURFER_HOME", ""):
                     exe = os.environ["FREESURFER_HOME"] + "/bin/mri_coreg"
             else:
                     raise RuntimeError("Could not find mri_coreg, source FreeSurfer or set the  FREESURFER_HOME environment variable")
        args = [exe, "--mov", t2_path, "--targ", t1_path, "--reg", lta_path]
        LOGGER.info("Running " + " ".join(args))
        retval = Popen(args).finish()
        if retval.retcode != 0:
              LOGGER.error(f"mri_coreg failed with error code {retval.retcode}. ")
              raise RuntimeError("mri_coreg failed registration")
```
        


You cant really benefit from the real utility of Popen here or parallel execution, such as iterating through stdout of the child process, but this is also a bit less error-prone.

dkuegler reviewed (commented):
I went through the code and have some comments.

Some of the points can probably be done by someone else, but there are a couple of questions we want to fix soon, before you invest much more work. Specifically, the questions regarding bias field correction.

santiestrada32 committed:
clean pipeline, comments, conflicts, functions, etc
add t2 option
select automatically processing mode
re-strucuture pre-processing
remove bias field correction
organize bin additions alphabetically
Modifications based on merge comments
Reimplement run_cmd
Change run_cmd for Popen from FastSurferCNN.utils.run_tools
Change config files from json to yaml

dkuegler commented on HypVINN/utils/visualization_utils.py:
```diff
@@ -161,10 +161,11 @@ def plot_qc_images(save_dir,orig_path, prediction_path, padd=45, lut_file=HYPVIN
 
     image = nib.as_closest_canonical(nib.load(orig_path))
     pred  = nib.as_closest_canonical(nib.load(prediction_path))
+    pred_arr = hypo_map_subseg_2_fsseg(pred.get_fdata(),reverse=True)
```
Instead of IMG.get_fdata(), you may want to use np.asarray(omg.datsobj) as the former always returns float arrays, but here you are loading a segmentation, i.e. int data.

santiestrada32 committed:
Add Snapshot as optional argument

santiestrada32 committed:
update modality selection mode

santiestrada32 committed:
Add documentation and simplify comparison code

santiestrada32 committed:
Add warning if resolution on multi-modal mode are different during registration

santiestrada32 committed:
add option for T2 image

santiestrada32 committed:
fix label upload

santiestrada32 committed:
TODO for integrating multi-modal in FastSurfer run

santiestrada32 committed:
Update readme file

santiestrada32 committed:
fix typos

santiestrada32 committed:
include TODOs for inference optimization

santiestrada32 committed:
update Hypvinn to FreeSurfer LUT

santiestrada32 committed:
Add documentation

36000 committed:
Update download_checkpoints.py
update network name
update download_checkpoints

santiestrada32 committed:
fix mapping

m-reuter commented:
please don't merge from dev, do a rebase instead. Rebase - of course - should not be done if we are officially collaborating on a branch it is checked out by other people. but if this is mainly on your fork, just rebase, which will add the dev changes before your changes and keeps the history linear.

santiestrada32 committed:
update log folder name

santiestrada32 committed:
update hypvinn information

santiestrada32 committed:
Fix ras headers issues when passing multi-modality data

santiestrada32 committed:
fix t2 path typo

santiestrada32 committed:
reduce the times of loading the orig image

dkuegler commented on HypVINN/README.md:
```diff
@@ -3,36 +3,32 @@
-Same as FastSurfer
-FreeSurfer should also be source to run the code and the mri_coreg ang mri_vol2vol binary should be also available
+* Same as FastSurfer.
+* If the T1w and T2w images are available and not co-registered, FreeSurfer should be source to run the registration code, and the mri_coreg and mri_vol2vol binary should also be available.
```
```suggestion
* If the T1w and T2w images are available and not co-registered, FreeSurfer should be sourced to run the registration code, and the mri_coreg and mri_vol2vol binaries should also be available.
```

dkuegler commented on HypVINN/README.md:
```diff
@@ -54,7 +50,7 @@ Run the HypVINN/run_pipeline.py which has the following arguments:
 The Hypothalamus pipeline can be run by using a T1 a T2 or both images. 
 Is recommended that all input images are bias field corrected and when passing both T1 and T2 they need to be co-registered.
-The pipeline can do all pre-processing by itself (steps 1 and 2) or omit this step if images are already curated beforehand.
+The pipeline can do all pre-processing by itself (steps 1) or omit this step if images are already curated beforehand.
```
```suggestion
The pipeline can do all pre-processing by itself (step 1). This step can be skipped if images are already registered externally. Note, that images are conformed as a first step, which can lead to additional interpolation reducing quality.
```

@santiestrada32 does the registration as part of the pipeline register to the conformed t1? Is there a way we can further resuce the interpolations? 

We don't have to include this in this PR though...

dkuegler commented on HypVINN/README.md:
````diff
@@ -65,33 +61,33 @@ The pipeline can do all pre-processing by itself (steps 1 and 2) or omit this st
    ```
 
-3. Run pipeline with no pre-processing
+3. Run pipeline with no registration step
````
```suggestion
3. Run pipeline without the registration step
```

dkuegler reviewed (commented):
This review was "overdue". Now it is down to me to fix run_fastsurfer.sh.

dkuegler committed:
Apply suggestions from code review

dkuegler committed:
update run_fastsurfer.sh for hypvinn
formatting

santiestrada32 committed:
rename third-ventricle to resolve freesurfer LUT integration issues

dkuegler committed:
update run_fastsurfer.sh for hypvinn
formatting

dkuegler committed:
Fix output file documentation
Add hypvinn url file

dkuegler committed:
doc/overview/OUTPUT_FILES.md
- remove/rename changed filenames
- format table

HypVINN/config/checkpoint_paths.yaml
- add config

HypVINN/data_loader/data_utils.py
- fix typing and formatting

HypVINN/utils/checkpoint.py
- fix YAML_DEFAULT

HypVINN/utils/mode_config.py
- set default values for get_hypvinn_mode

HypVINN/inference.py
- fix inclusion of ModalityMode

HypVINN/run_prediction.py
- move HypVINN/run_pipeline.py into run_prediction.py
- fix typing, e.g. FileBasedHeader
- fix function parameters
- add help text to hypo_segfile argument
- fix passing of t1_path and t2_path
- various other changes

m-reuter commented on run_fastsurfer.sh:
```diff
@@ -875,6 +886,23 @@ if [[ "$run_seg_pipeline" == "1" ]]
+        then
+          # ... we have a t2 image, bias field-correct it
+        echo "INFO: Running N4 bias-field correction of the t2" | tee -a "$seg_log"
+        cmd=($python "${reconsurfdir}/N4_bias_correct.py" "--in" "$t2"
```
Have you ever tested our N4 wrapper script for the T2 image?

dkuegler replied:
@santiestrada32 has used the script for this (we even found a bug in there doing that). But this PR also still needs some testing anyways...

dkuegler committed:
Add HypVINN "basic" api docfiles
Fix typo in N4_bias_correct.py

dkuegler committed:
- Fix errors in t1_only mode
- Fix errors in t1t2 mode
- Fix qc_snaps argument in run_fastsurfer.sh
- Fix reg_mode test in run_fastsurfer.sh

dkuegler committed:
remove run_pipeline from doc

santiestrada32 committed:
Update documentation for HypVINN module

santiestrada32 committed:
hypvinn config file v1.1

santiestrada32 committed:
Updated model weights

update model weight (version 1.1) for hypvinn
update download script

santiestrada32 committed:
Better input handling of the input T2 file

Convert the T2 image to unit8 to save hard disc space for registration
- set --uchar for N4_bias_correct.py
- `conform.py --no_force_(img_size|vox_size|lia) --t2 <arg>` if --no_biasfield
Copy the T2 file to $SUBJECTS_DIR/$SID/mri/orig/T2.001.mgz

taha-abdullah committed:
Documentation for HypVINN:

- Added docstring for functions in HypVINN

taha-abdullah committed:
sphinx syntax issues:

- adding space before colon separating parameter name and type

dkuegler committed:
Apply suggestions from code review

taha-abdullah committed:
Applying changes requested in code review

santiestrada32 commented:
I think the branch is ready to merge. Or something else needs to be tested?

dkuegler commented:
I think there is one issue left that I need to resolve, which is based around conforming/pre-processing of the T2 image.

In the current code, the conversion to uchar is independent of the bias field correction. This is not necessary, the bias field script has a `--uchar` flag that actually incorporates this functionality exactly.

The next/other open issue is the processing if the `--t2` image is given, but `--no_biasfield` is specified. This would mean, that the biasfield has been computed externally. That would be fine, but the conversion to uchar (and robust rescaling) is not ensured within the pipeline. This change needs a small adaptation of the conform script, I believe to also enable robust rescaling without reorientation, which is a feature that we also regularly need in other contexts.

I have a branch where both of these issues are started (actually 1 should be finished, 2 is started but not tested yet).

Finally, there used to be a sphinx documentation problem, that I was supposed to fix, but did not get around to yet. So both are not really critical issues, but more "side-issues".

santiestrada32 commented:
I think it would be nice to merge my Branch to dev before OHBM ( Monday, June 24), as I will be presenting HypVINN.

In the preproc.py, before I registered T1 and T2, I ensured that the T2 image was converted to uchar and that robust rescaling was applied. I don't do this if they put none as a registration step, so I agree that improving the T2 conform will be nice.

Let me if I can help with something

m-reuter commented:
We can also merge now and do the rest as separate PRs?

dkuegler committed:
Add mask_file as a parameter (with default value) for saving instead of a constant
Make saving a main thread function instead of a future/task

dkuegler committed:
Refactor and cleanup of hypvinn

- rename mode "multi" to "t1t2"
- reformatting for line length
- replace single quotes to double quotes
- add typing information
- clean up docstrings
- replace os.path with pathlib.Path

dkuegler committed:
run_fastsurfer.sh
- Fix for bash-3.2 (|& => 2&>1 |)
- Add warning if T2 is passed, but biasfield is skipped --no_biasfield

dkuegler committed:
Documentation

- add comment paper vs. retrained weights.
- add hypothal module in main README
- fix sphinx errors
- add an additional message in fix_links to debug errors
- add the hypvinn script documentation

dkuegler commented:
I am going to merge this now. This should be ready for the base use case, but I want to fix the robust scaling of the T2 image. This can be a different PR though.
